### DISCUSSION TITLE
Help with customizing .src files

### DISCUSSION THREAD

**Discussion by ElysiumFic** [seanap/Plex-Audiobook-Guide#78] [General]
Not actually an issue, just a question...

I'm a little nitpicky in the way I like my audiobooks sorted and streamed using Prologue, and also I like to populate some metadata that Plex doesn't use so I won't every have to redo all my tagging in the event that some day another way of managing my library comes along that I like better.

I'm fairly decent at picking up elementary-level code via simple monkey-see/monkey-do, so I've _mostly_ been able to tweak the .src files for scraping Audible info so that it suits my particular preferences.

However, I apparently have a mental block when it comes to regex stuff. I've tried many times over the last five years to try to wrap my head around it, but it just..._doesn't_ compute for me.

So I'm able to, for example, remove things like "(Unabridged)" at the end of a title and the (usually unnecessary if not downright redundant) "Series" at the end of a series title, and remove the # from the Subtitle/ContentGroup fields ", Book #" just because find it more aesthetically pleasing.

There are two things I would like to do, however, that I can't figure out how to automate upon import using the .src script.

The first is automatically populate "Sort Artist/AlbumArtist/Composer" with Last Name, First formatting, because I can't work out the syntax for separating the name string at the final space in the string, then swap the two with a comma in between (which isn't a bulletproof solution, I know, but it handles ~90% of names.)

I also like to separate multiple authors and narrators by semicolons, rather than commas. This makes the delimiter easier to locate in situations where you have "Author Name, Ph.D." or whatever. Admittedly, the main reason I hope to have this is because someday, I'm hoping to finish the library database app I'm writing for MacOS/iOS, which would give each individual author their own record in the database instead of lumping multiple authors together. (So, for instance, books by "Robert Jordan; Brandon Sanderson" could be found when you look up either author's bibliography.) But that's another endeavor.

I would also like to import the Publisher's Summary to UNSYNCEDLYRICS (or long description, but most people seem to use the lyrics tag for this) complete with paragraph breaks (from there, it can be filtered/shortened further for use in the shorter COMMENT/DESCRIPTION tags.) The `<p> </p>` html tags are filtered out in the regex here, I presume?

```
if "bc-text-bold\" >Publisher's Summary</h2>"
	outputto "UnsyncedLyrics"
	findline "<span class"
	joinuntil "</span>"
	regexpreplace "</?[^><]+>" ""
	unspace
	regexpreplace "  +" " "
	replace "bc-color-secondary\" >" ""
	sayrest
else
```

But I'm not sure where. I know the `unspace` is used to truncate leading and trailing whitespaces. Does that include paragraph breaks? 

I'd also like to trim leading "The"/"A"/"An" etc from the album/title sorting.

I can probably do my monkey-see/monkey-do stuff if I can see samples of code where this is done, so there isn't a need to do it for me, unless it's quicker and easier to just demonstrate on the fly?

**Comment by seanap** [comment]
Full disclosure, I am not a programmer.  The majority of all my scripts are the result of hrs and hrs of trial and error, googling, and stackoverflow.

The .src script is an html scrapper, so when it comes to modifying the results of what it scrapes it can be very lacking in functionality.  It looks at the html source code of a web page, it helped me to imagine that we are trying to tell the computer to place a virtual curser in the exact place we want so we can copy/paste.  We have to tell it where to start looking and where to end, then we can use regex to remove html syntax, then we write whatever is left to a tag.  The unspace command is used if our selection spans multiple lines, in your example above it has to findline xxx and join all the following lines until yyy, the unspace command essentially trims all the tab spacing so the html formatting doesn't effect our output.

#1 
To change the formatting of the artists names it would be a whole lot easier to add an action to your action command that you run after scrapping the tags. Something like this for each field:
```
Action: Format Value
Field: AlbumArtist
String: $regexp(%albumartist%,(.) (.),$2',' $1)
```

#2 
This question can get a little tricky. Do you want the artist tag to be "first last; first last" or "Last, first; Last, First"? For the first-last case then this is as simple as changing line 326 in the src script from a comma to a semicolon, eg:
```
# Set Artist = artist, narrator
outputto "artist"
sayoutput "Albumartist"
say "; "
sayoutput "Composer"
```
For the last-first case then we probably would want to remove that section from the scr script altogether and instead use an action that would run after the Last-first action above to combine those into the artist tag.  Between using the mp3tag action gui and searching the mp3tag forums you can do just about anything you want.

#3
The Publisher's Summary is currently being set in the COMMENT and DESCRIPTION tags.  The reason for setting to both is because Mp3's use the comment tag, and M4B's use the Description tag.  To add the summary to UNSYNCEDLYRICS add the following below line 337:
```
# Set Comment to UNSYNCEDLYRICS
outputto "UNSYNCEDLYRICS"
sayoutput "Comment"
```

#4
To move title articles to the end it would probably be best to also have this as an action.  Here is the regex to move "The" to the end from https://community.mp3tag.de/t/regular-expressions/521/15:
```
Regular expression: ^The (.+)
Replace with: $1, The
```

**Reply by ElysiumFic** [reply]
> Full disclosure, I am not a programmer. The majority of all my scripts are the result of hrs and hrs of trial and error, googling, and stackoverflow.
> 

I get that. I recently spent two years teaching myself Swift with the help of a mentor, but before that, the last time I'd done any coding beyond very elementary HTML/XML was in 1988 when I took BASIC in high school.

> The .src script is an html scrapper, so when it comes to modifying the results of what it scrapes it can be very lacking in functionality. It looks at the html source code of a web page, it helped me to imagine that we are trying to tell the computer to place a virtual curser in the exact place we want so we can copy/paste. We have to tell it where to start looking and where to end, then we can use regex to remove html syntax, then we write whatever is left to a tag. The unspace command is used if our selection spans multiple lines, in your example above it has to findline xxx and join all the following lines until yyy, the unspace command essentially trims all the tab spacing so the html formatting doesn't effect our output.
> 

Yes, I got that general gist by looking over what you had done, but I don't even know what language the .src file code is, or what some of the command you're using are. That said, if I were doing this in Swift, I would probably parse the HTML much the same way.

> #1 To change the formatting of the artists names it would be a whole lot easier to add an action to your action command that you run after scrapping the tags. Something like this for each field:
> 
> ```
> Action: Format Value
> Field: AlbumArtist
> String: $regexp(%albumartist%,(.) (.),$2',' $1)
> ```
I have been trying to do a lot of this in actions, working with what the .src has already parsed into the various tags. But it seems like the options for implementing more sophisticated or intricate searches and replacements is limited in actions, since you're restricted to one line of commands, right?

Since it will be better for me going forward not just to copy this code, but to understand what it's doing so I can adapt things as necessary, I'm going to try to figure out what this is doing.

I would actually be doing this in the "ArtistSort" or "AlbumArtistSort" tags. So what I would want to do is format those tags, copy the string from "Artist"/"AlbumArtist" and then adapt it as necessary.

I know the `.` is the regex wildcard character. I'm still not entirely sure I understand what parentheses do, particularly the second point listed in the [MP3Tag documentation](https://docs.mp3tag.de/regex/#parenthesis). It looks like it's indicating that the search will look for any two strings separated by a space? And somehow those two strings are designated `$1` and `$2`, so `$2',' $1` reverses the order of the strings, separated by a comma?

Which might be problematic when dealing with names that are comprised of initials that might have spaces between them (eg J. K.. Rowling) or whatever, or three-word names (eg Edgar Allen Poe).

> 
> #2 This question can get a little tricky. Do you want the artist tag to be "first last; first last" or "Last, first; Last, First"? For the first-last case then this is as simple as changing line 326 in the src script from a comma to a semicolon, eg:
> 
> ```
> # Set Artist = artist, narrator
> outputto "artist"
> sayoutput "Albumartist"
> say "; "
> sayoutput "Composer"
> ```
> 

Ideally, I would want "AlbumArtist" to be "First1 Last1; First2 Last2" (and same for narrator) and then "(Album)ArtistSort" would be "Last1, First1; Last2, First2"

> For the last-first case then we probably would want to remove that section from the scr script altogether and instead use an action that would run after the Last-first action above to combine those into the artist tag. Between using the mp3tag action gui and searching the mp3tag forums you can do just about anything you want.
> 
Yeah, like I said, I've been experimenting with this. And I've found it's sometimes faster and easier to edit the .mta files in Notepad++ rather than in the gui. It's just the regex that is tripping me up, because I don't generally get it.

> #3 The Publisher's Summary is currently being set in the COMMENT and DESCRIPTION tags. The reason for setting to both is because Mp3's use the comment tag, and M4B's use the Description tag. To add the summary to UNSYNCEDLYRICS add the following below line 337:
> 
> ```
> # Set Comment to UNSYNCEDLYRICS
> outputto "UNSYNCEDLYRICS"
> sayoutput "Comment"
> ```
> 
I know I can populate "UnsyncedLyrics" by copying from one of the other fields, but what I would actually like to do is import to UnsyncedLyrics in such a way that the line breaks in the publisher summary aren't removed.

> #4 To move title articles to the end it would probably be best to also have this as an action. Here is the regex to move "The" to the end from https://community.mp3tag.de/t/regular-expressions/521/15:
> 
> ```
> Regular expression: ^The (.+)
> Replace with: $1, The
> ```

Thank you. I think I get what the code is doing there.I just need to look up some of the symbols (again) to make sure I understand what they're for.

**Reply by seanap** [reply]
I believe the Web Source Framework was created by the creator of mp3tag, it's kinda like a proprietary language. Also, I believe the regex only applies to the parser. You can't regex tag values, only the source code.

Regarding # 4 if unspace is removing the paragraph lines then you can try removing the unspace but if that doesn't work the only way I know of to fix that is to open the audible source code and find where the publishers summary is and use a better `FINDLINE "bc-text-bold\" >Publisher's Summary</h2>"` to something that wouldn't need unspace. The summary was a tough nut to crack and there are a lot of edge cases that I had to account for since audibles source code is super inconsistent.

Web scraping has a lot of drawbacks and limitations, especially when using the mp3tag parser framework. Using the audnexus plugin for plex allows you to select Last, First for your authors and it pulls the description properly formatted so the front end looks perfect even if the tags are so-so. What we need is an audnexus tagger so we can use the audiobook database via api.
